src/dispatching/updaters.rs

```diff
@@ -0,0 +1,171 @@
+//!      |           timeout          |
+//!      ↓                            ↓
+//! Yes  |-------[updates 2..5]------>|
+//!      ↑                            ↑
+//!      |           delay            |
+//!      ↓                            ↓
+//!      |<-------[offset = 5]--------| Updates?
+//!      ↑                            ↑
+//!      |           timeout          |
+//!      ↓                            ↓
+//! Nope |--------------------------->|
+//!      |                            |
+//!      ~    and so on, and so on    ~
+//! </pre>
+//!
+//! <a id="1" href="#1b">^1</a> A timeout can be even 0
+//!   (this is also called short polling),
+//!   but you should use it **only** for testing purposes.
+//!
+//! <a id="2" href="#2b">^2</a> Large delays will cause in bot lags,
+//!   so delay shouldn't exceed second.
+//!
+//! <a id="3" href="#3b">^3</a> Note that if Telegram already have updates for
+//!   you it will answer you **without** waiting for a timeout.
+//!
+//! <a id="4" href="#4b">^4</a> `offset = N` means that we've already received
+//!   updates `0..=N`.
+//!
+//! [`Updater`]: Updater
+//! [`polling_basic`]: polling_basic
+//! [`polling_advanced`]: polling_advanced
+//! [`Dispatcher`]: crate::dispatching::Dispatcher::dispatch
+//! [`Box::get_updates`]: crate::Bot::get_updates
+//! [getting updates]: https://core.telegram.org/bots/api#getting-updates
+//! [long]: https://en.wikipedia.org/wiki/Push_technology#Long_polling
+//! [short]: https://en.wikipedia.org/wiki/Polling_(computer_science)
+//! [webhook]: https://en.wikipedia.org/wiki/Webhook
+
+use futures::{stream, Stream, StreamExt};
+
+use crate::{
+    bot::Bot, requests::payloads::AllowedUpdate, types::Update, RequestError,
+};
+use std::{convert::TryInto, time::Duration};
+
+/// A generic updater.
+pub trait Updater<E>: Stream<Item = Result<Update, E>> {
+    // TODO: add some methods here (.shutdown(), etc).
+}
+impl<S, E> Updater<E> for S where S: Stream<Item = Result<Update, E>> {}
+
+/// Returns a long polling updater with the default configuration.
+///
+/// It is the same as calling [`polling_advanced`] with `timeout` of 30 seconds,
+/// `limit=100` and receive all kinds of updates.
+///
+/// [`polling_advanced`]: polling_advanced
+pub fn polling_basic(bot: &Bot) -> impl Updater<RequestError> + '_ {
+    polling_advanced::<&[_]>(bot, Duration::from_secs(30), 100, &[])
+}
+
+/// Returns a long/short polling updater with some additional options.
+///
+/// - `bot`: Using this bot, the returned updater will receive updates.
+/// - `timeout`: A timeout for polling.
+/// - `limit`: Limits the number of updates to be retrieved at once. Values
+///   between 1—100 are accepted.
+/// - `allowed_updates`: A list the types of updates you want to receive.
+pub fn polling_advanced<'a, A>(
+    bot: &'a Bot,
+    timeout: Duration,
+    limit: u8,
+    allowed_updates: A,
+) -> impl Updater<RequestError> + 'a
+where
+    A: Into<&'a [AllowedUpdate]>,
```

Comment:
Em, I don't think `Into<&'a [AllowedUpdate]>` is a thing, that anyone ever wanted... It isn't even implemented for `Vec<_>`! Just accept `Vec<_>` and don't mind.

Reply:
Did you tried it? I came up with the error:

```
cannot move out of `allowed_updates`, a captured variable in an `FnMut` closure
```